### PR TITLE
Threading

### PR DESCRIPTION
**Pull request by chuyskywalker**
Improve overall thread handling
- Prevent flask from re-running the app causing duplicate search threads 
  - _(Technically this disables hot-reloading of code, if I understand correctly, but that doesn't seem a high cost. Obliterating niantic servers does.)_
- Prevent scan system from recursing to infinity
- Increase the search loop default delay (with the search threading, give Niantic's poor servers breathing room)
- Handle the mock data/search thread duality better
- Make `-ns` mode interruptable

Misc fixes found while working:
- Fix EOL on config.ini
- Ignore PEP8 harder
- Allow enough space for "connectionpool" debug output

## Motivation and Context
Discussion here: https://github.com/AHAAAAAAA/PokemonGo-Map/commit/3bafbfdfd5f219b87a8c127cea433d1a83860cc5

## How Has This Been Tested?
Run under `-ns` and `-m` modes combinations.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation. _**N/A**_
- [x] I have updated the documentation accordingly. _**N/A**_

### CODE REVIEW

**Comment by AHAAAAAAA** [comment]
Traceback (most recent call last):
  File "runserver.py", line 71, in <module>
    search_thread = start_locator_thread(args)
  File "runserver.py", line 29, in start_locator_thread
    search_thread = Thread(target=fake_search_loop, args=(args,))
NameError: global name 'fake_search_loop' is not defined

**Comment by chuyskywalker** [comment]
Thanks, I'll figure out what's going on and get the conflicts resolved.

**Comment by chuyskywalker** [comment]
All better now.

**Comment by JZ6** [comment]
+1

File: pogom/models.py [hunk 1]
```diff
@@ -15,8 +15,6 @@
 from .transform import transform_from_wgs_to_gcj
 from .customLog import printPokemon
 
-logging.basicConfig(level=logging.INFO, format='%(asctime)s [%(module)11s] [%(levelname)7s] %(message)s')
-
```
Comment (chuyskywalker) [inline pogom/models.py]: This should not be redefined all over the damn place.

File: pogom/models.py [hunk 2]
```diff
@@ -27,7 +25,6 @@ def init_database():
     if db is not None:
         return db
 
-    print args.db_type
```
Comment (chuyskywalker) [inline pogom/models.py]: Derp

File: pogom/search.py [hunk 3]
```diff
@@ -1,9 +1,21 @@
...
+Search Architecture:
+ - Create a Queue
+   - Holds a list of locations to scan
+ - Create N search threads
+   - Each search thread will be responsible for hitting the API for a given scan location
+ - Create a "overseer" loop
+   - Creates/updates the search grid, populates the Queue, and waits for the current search itteration to complete
+   -
+'''
+
```
Comment (chuyskywalker) [inline pogom/search.py]: Some helper text to introduce people to WTF is going on here

File: pogom/search.py [hunk 4]
```diff
@@ -19,16 +31,16 @@
...
 lat_gap_meters = 150
 lng_gap_meters = 86.6
 
-#111111m is approx 1 degree Lat, which is close enough for this
+# 111111m is approx 1 degree Lat, which is close enough for this
 meters_per_degree = 111111
 lat_gap_degrees = float(lat_gap_meters) / meters_per_degree
 
-search_queue = Queue(config['SEARCH_QUEUE_DEPTH'])
+search_queue = Queue()
```
Comment (chuyskywalker) [inline pogom/search.py]: We need not artificailly limit the queue size -- control that through the thread count (which we already do)

File: pogom/search.py [hunk 5]
```diff
@@ -123,32 +149,55 @@ def search_thread(args):
...
             else:
-                log.info('Map Download failed. Trying again.')
+                log.info('Map download failed, waiting and retrying')
+                log.debug('{}: itteration {} step {} failed'.format(threadname, i, step))
                 time.sleep(config['REQ_SLEEP'])
 
         time.sleep(config['REQ_SLEEP'])
+        q.task_done()
 
-def process_search_threads(search_threads, curr_steps, total_steps):
```
Comment (chuyskywalker) [inline pogom/search.py]: Completely unused

File: pogom/utils.py [hunk 6]
```diff
@@ -38,7 +37,9 @@ def get_args():
     parser.add_argument('-p', '--password', help='Password')
     parser.add_argument('-l', '--location', type=parse_unicode, help='Location, can be an address or coordinates')
     parser.add_argument('-st', '--step-limit', help='Steps', type=int, default=12)
-    parser.add_argument('-sd', '--scan-delay', help='Time delay before beginning new scan', type=float, default=1)
+    parser.add_argument('-sd', '--scan-delay', help='Time delay between requests in scan threads', type=float, default=5)
+    parser.add_argument('-td', '--thread-delay', help='Time delay between each scan thread loop', type=float, default=5)
+    parser.add_argument('-ld', '--login-delay', help='Time delay between each login attempt', type=float, default=5)
```
Comment (chuyskywalker) [inline pogom/utils.py]: Split all these up for better control over the niggly details. A single global "delay" doesn't make sense. If you run with 10 threads, you should probably pull the horses back a bit on each thread, but the wait time between loops shouldn't also get delayed. Having a login specific delay is also nice so you can set it a bit higher/lower.

File: runserver.py [hunk 7]
```diff
@@ -12,37 +12,34 @@
...
+log = logging.getLogger()
 
 if __name__ == '__main__':
-    logging.basicConfig(level=logging.INFO, format='%(asctime)s [%(module)11s] [%(levelname)7s] %(message)s')
+    args = get_args()
+
+    if args.debug:
+        log.setLevel(logging.DEBUG);
+    else:
+        log.setLevel(logging.INFO);
```
Comment (chuyskywalker) [inline runserver.py]: Doing it this way means we don't have to put EVERYTHING in the if statement further on. They will inherit.

**Comment by chuyskywalker** [comment]
I've now tested this with

```
rm pogom.db; python ./runserver.py <keys, logins, etc> -l "40.770045 -73.975167" -st 3 -t 1 -ld 15 -td 5 -sd 0
```

Which emulates a slow, single threaded search and got this map (central park):

2016-07-26 22_08_07-pokemon go map

Then again like:

```
rm pogom.db; python ./runserver.py <keys, logins, etc> -l "40.770045 -73.975167" -st 3 -t 10 -ld 15 -td 15 -sd 0
```

Which is, effectively, 10 threads as fast as possible with a delay between full loops, which netted me:

2016-07-26 22_10_50-pokemon go map

Overall, I think this looks pretty good.

**Comment by AHAAAAAAA** [comment]
Tested, agree with the separation of delays. +1, documentation for Usage in Wiki must be updated.

**Comment by chuyskywalker** [comment]
Will update [mainpage](https://github.com/AHAAAAAAA/PokemonGo-Map/wiki) and [usage](https://github.com/AHAAAAAAA/PokemonGo-Map/wiki/Usage) page as needed after merge

**Comment by chuyskywalker** [comment]
Wiki's updated

**Comment by rneu31** [comment]
Did this commit break something? Rebased to origin/develop, Pokemon were not found. Reset --hard to the commit before this one and everything is working again. Will look into this further later.

**Pull request merged**